src/IO/ADIOS/ADIOS2IOHandler.cpp

```diff
@@ -20,58 +20,1225 @@
  */
 #include "openPMD/IO/ADIOS/ADIOS2IOHandler.hpp"
 
-#include <utility>
-
+#include "openPMD/Datatype.hpp"
+#include "openPMD/IO/ADIOS/ADIOS2FilePosition.hpp"
+#include "openPMD/IO/ADIOS/ADIOS2IOHandler.hpp"
+#include "openPMD/auxiliary/Environment.hpp"
+#include "openPMD/auxiliary/Filesystem.hpp"
+#include "openPMD/auxiliary/StringManip.hpp"
+#include <iostream>
+#include <string>
 
 namespace openPMD
 {
+#if openPMD_USE_VERIFY
+#define VERIFY( CONDITION, TEXT )                                              \
+    {                                                                          \
+        if ( !( CONDITION ) )                                                  \
+            throw std::runtime_error( ( TEXT ) );                              \
+    }
+#else
+#define VERIFY( CONDITION, TEXT )                                              \
+    do                                                                         \
+    {                                                                          \
+        (void)sizeof( CONDITION );                                             \
+    } while ( 0 );
+#endif
+
+#define VERIFY_ALWAYS( CONDITION, TEXT )                                       \
+    {                                                                          \
+        if ( !( CONDITION ) )                                                  \
+            throw std::runtime_error( ( TEXT ) );                              \
+    }
+
 #if openPMD_HAVE_ADIOS2
-ADIOS2IOHandler::ADIOS2IOHandler(std::string path, AccessType at)
-        : AbstractIOHandler(std::move(path), at),
-          m_impl{new ADIOS2IOHandlerImpl(this)}
-{ }
 
-ADIOS2IOHandler::~ADIOS2IOHandler()
-{ }
+#if openPMD_HAVE_MPI
 
-std::future< void >
-ADIOS2IOHandler::flush()
+ADIOS2IOHandlerImpl::ADIOS2IOHandlerImpl( AbstractIOHandler * handler,
+                                          MPI_Comm communicator )
+: AbstractIOHandlerImplCommon( handler ), m_comm{communicator},
+  m_ADIOS{communicator, ADIOS2_DEBUG_MODE}
 {
-    return m_impl->flush();
 }
 
-#endif
+#endif // openPMD_HAVE_MPI
 
-#if openPMD_HAVE_ADIOS2 && !openPMD_HAVE_MPI
-ADIOS2IOHandler::ADIOS2IOHandler(std::string path, AccessType at)
-        : AbstractIOHandler(std::move(path), at),
-          m_impl{new ADIOS2IOHandlerImpl(this)}
-{ }
+ADIOS2IOHandlerImpl::ADIOS2IOHandlerImpl( AbstractIOHandler * handler )
+: AbstractIOHandlerImplCommon( handler ), m_ADIOS{ADIOS2_DEBUG_MODE}
```

Comment:
Clang-tidy does complain about the `ADIOS2` constructor here.
  https://travis-ci.org/openPMD/openPMD-api/jobs/535234522

Maybe try:
```suggestion
: AbstractIOHandlerImplCommon( handler ),
#if openPMD_HAVE_MPI
    m_ADIOS{MPI_COMM_SELF, ADIOS2_DEBUG_MODE}
#else
    m_ADIOS{ADIOS2_DEBUG_MODE}
#endif
```
?